### ISSUE TITLE
Get a more precise timing for the stage execution time

### ISSUE THREAD

**Issue by dcoraboeuf** [jenkinsci/ontrack-plugin#54]
The current way to get the execution time for a stage if to get the total timing of the stage ; this includes queuing time and node preparation time.

We want the time to be _only_ about the execution of the stage.

**Comment by dcoraboeuf** [comment]
In a declarative pipeline, when using:

```
stage("x") {
   agent { }
}
```

the time to create the agent if not available is taken into account when measuring the timing of the `x` stage:

<img width="603" alt="image" src="https://user-images.githubusercontent.com/1206964/53976881-2e123600-4108-11e9-89c2-4a1873c19e29.png">

It's a same issue with a scripted pipeline written this way:

```
stage("x") {
   node {
   }
}
```

<img width="460" alt="image" src="https://user-images.githubusercontent.com/1206964/53977718-f0aea800-4109-11e9-8d75-b3aaeb0e802e.png">

When inverting the syntax:

```
node {
   stage("x") {}
}
```

then the computation is OK:

<img width="435" alt="image" src="https://user-images.githubusercontent.com/1206964/53978126-c4475b80-410a-11e9-9c56-c1b01431e367.png">

So it means that have to take into account the cases of:

i) declarative pipeline
ii) node creation inside a declarative stage

The case iii) node creation _outside_ of a stage is already correct.

**Comment by dcoraboeuf** [comment]
Proposal: when computing the execution time of a stage, we go up the step tree until we find the stage step itself. We take its execution time `s`.

On the way up the tree, we collect each "Allocate node" step execution time into the `n` time.

The computed stage execution time will be `x = s - t`.

This will solve cases i) and ii) (and also work for iii) where `t = 0`)

**Comment by dcoraboeuf** [comment]
"Allocate Node: Start" (`n`) and "Allocate Node: Body: Start" (`b`) have to be taken into account.

If no such step is seen before we get to the stage, `t = s`.

If not, `t = sum(b)` where `b` is the list of "Node: Body: Start" steps.

**Comment by dcoraboeuf** [comment]
Fixed in 3.4
